### Instructions
Pick an algorithm from the [Wikipedia sort page](https://en.wikipedia.org/wiki/Sorting_algorithm) that hasn't been covered in the lecture (we covered insertion, merge, and default sorts) and implement it in Python.   Good sorts to try are: heap, selection, and quick sort.  

In addition, see if you can figure out why the chosen sort runs faster or slower than other sorting algorithms.

In [1]:
import time
import random
import numpy as np

# Set seed.
random.seed(a=42)

# Create our default list.
short_list = list(random.sample(range(1000000), 10))
long_list = list(random.sample(range(1000000), 10000))

### Quicksort
I've decided to implement quicksort.  With quicksort, the idea is that you have an array of numbers, and you pick a number to be a pivot.  Each number in the list is sorted so that any number smaller than the pivot is placed to the left of the pivot, and any number larger than the pivot is placed to the right; this gives us left and right partitions.  

Picking a good pivot is essential as you want the pivot to be a number near the middle of the list.  People often pick the pivot to be the first or last number in the list, but this can sometimes be problematic if the list is already sorted for example.  

Another method is to take the median of three numbers, the first, last, and middle value, which is what I'll do with the get_pivot function.

In [2]:
def get_pivot(arr):
    low = arr[0]
    high = arr[-1]
    mid = arr[len(arr)//2]
    median = np.median([low, mid, high])
    if median == arr[0]:
        return 0
    elif median == arr[-1]:
        return len(arr)-1
    else:
        return len(arr)//2

How do we get the partitions?  First we put the pivot value at the beginning of the array.  Then the "border" will start in position 1 (since python is 0 index based).  The border is what creates the partition.  We work our way down the list and everything with a value lower than the pivot value will be put in a list called left, and everything with a value higher than the pivot value will be put in a list called right.  We recursively return the value of the quicksorted left, pivot value, and quicksorted right lists. 

In [3]:
def quicksort(arr):
    # if arr has more than 1 element...
    if len(arr) > 1:
        # get pivot number and index
        pivotIndex = get_pivot(arr)
        pivotNum = arr[pivotIndex]

        low = 0 #lowest element is always index 0
        
        # pivot number goes to beginning of array (position 0)
        arr[pivotIndex], arr[low] = arr[low], arr[pivotIndex]

        # border to position 1
        borderIndex = 1

        for i in range(borderIndex, len(arr)):
            if arr[i] < pivotNum:
                arr[borderIndex], arr[i] = arr[i], arr[borderIndex]
                borderIndex += 1
        arr[borderIndex-1], arr[0] = arr[0], arr[borderIndex-1]
        return quicksort(arr[:borderIndex-1]) +  [pivotNum] + quicksort(arr[borderIndex:])
    else:
        return arr

As a test, I will see how long it takes to sort the short list and long list using the quicksort method I wrote...I get 0.001 and 0.44 seconds respectively.

In [4]:
start_time = time.time()
quicksort(short_list)
print('{} seconds'.format(time.time() - start_time))

0.0009984970092773438 seconds


In [5]:
start_time = time.time()
quicksort(long_list)
print('{} seconds'.format(time.time() - start_time))

0.4375779628753662 seconds


What about if I did this for a list that is already sorted?  The short list is about the same speed, but the long list takes longer than the randomized list.  This is most likely due to the function needing to switch elements more often.

In [6]:
start_time = time.time()
quicksort(np.arange(10))
print('{} seconds'.format(time.time() - start_time))

0.0010001659393310547 seconds


In [7]:
start_time = time.time()
quicksort(np.arange(10000))
print('{} seconds'.format(time.time() - start_time))

0.5514869689941406 seconds


As a sanity check, this type of sort is bad if the pivot point is chosen incorrectly.  The worst case scenario would be choosing the pivot point as the first or last element in a sorted list. I'll test that out now and summarize all 8 results in the table below.

| List Type | Time (sec)   | Time of Non-Optimizal Pivot (sec)
|------|------|
|   Short List  | 0.001| 0.003 |
|   Long List  | 0.438| 0.667|
|   Sorted Short List  | 0.001|0.001|
|   Sorted Long List  | 0.551|0.790|

It is pretty clear in the long list that choosing the correct pivot point is important, as the sorting takes ~0.15 seconds longer than the more optimal pivot point.  

In [8]:
def quicksort2(arr):
    # if arr has more than 1 element...
    if len(arr) > 1:
        # get pivot number and index
        pivotIndex = 0
        pivotNum = arr[pivotIndex]

        low = 0 #lowest element is always index 0
        
        # pivot number goes to beginning of array (position 0)
        arr[pivotIndex], arr[low] = arr[low], arr[pivotIndex]

        # border to position 1
        borderIndex = 1

        for i in range(borderIndex, len(arr)):
            if arr[i] < pivotNum:
                arr[borderIndex], arr[i] = arr[i], arr[borderIndex]
                borderIndex += 1
        arr[borderIndex-1], arr[0] = arr[0], arr[borderIndex-1]
        return quicksort(arr[:borderIndex-1]) +  [pivotNum] + quicksort(arr[borderIndex:])
    else:
        return arr

In [9]:
start_time = time.time()
quicksort2(short_list)
print('{} seconds'.format(time.time() - start_time))

0.003004789352416992 seconds


In [10]:
start_time = time.time()
quicksort2(long_list)
print('{} seconds'.format(time.time() - start_time))

0.6673743724822998 seconds


In [11]:
start_time = time.time()
quicksort2(np.arange(10))
print('{} seconds'.format(time.time() - start_time))

0.0009982585906982422 seconds


In [12]:
start_time = time.time()
quicksort2(np.arange(10000))
print('{} seconds'.format(time.time() - start_time))

0.7902567386627197 seconds
